File: Java/PackageDeliverySystem.java
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import java.util.List;

public class PackageDeliverySystem {
    public static void main(String[] args) throws IOException {

        //input file reading
        Scanner scan = new Scanner(System.in);
        System.out.println("Enter the name of file with packages: ");
        String fileName = scan.nextLine();
        String deliveryFile ="deliveries.txt";
        String errorLog = "log.txt";

        //creating/manging collection of package objects

        //to count how many packages are in the text file
        int numPackages = 0;
        try(BufferedReader br = new BufferedReader(new FileReader(fileName))){
            while((br.readLine()) != null) {
                numPackages++;
            }
        } catch (IOException e) {
            BufferedWriter writer = new BufferedWriter(new FileWriter(errorLog, true));
            writer.write("IOException");
            writer.newLine();;
            writer.close();
        }

        //reading file
        File file = new File(fileName);
        Scanner reader = new Scanner(file);
        ArrayList<String> list = new ArrayList<>();
        ArrayList<Pack> packs = new ArrayList<>();

        //Make an array of strings. Each element will be one line in the text file
        try {
            BufferedReader br = new BufferedReader(new FileReader(fileName));
            String line;
            while ((line = br.readLine()) != null) {
                list.add(line);
            }
            br.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

        //convert each string into a pack object and put in a pack array
        for (int i=0; i<numPackages; i++) {
            String[] data = list.get(i).split(",");
            if (data[0].equals("S") || data[0].equals("R")) {
                packs.add(parsePackData(data));
            } else {
                try {
                    BufferedWriter br = new BufferedWriter(new FileWriter(errorLog, true));
                    br.write("IOException\n");
                    br.close();;
                } catch (IOException ex) {
                    ex.printStackTrace();
                }
            }
        }

        /*
        make an ArrayList of packs
        when a line in the text file is incorrectly formatted,
        the packs array will make a pack that is null
        when we convert it to an ArrayList,
        it will be able to remove that null object
         */

        List<String> stringPacks = new ArrayList<>();
        for (Pack pack : packs) {
            stringPacks.add(pack.getDeliveryDate().toString());
        }

        List<String> days = new ArrayList<>();
        for(String str : stringPacks) {
            int indexSlash = str.indexOf("/");
            days.add(str.substring(indexSlash+1, indexSlash+3));
        }


        ArrayList<Integer> intDays = new ArrayList<>();
        for (String str : days) {
            intDays.add(Integer.parseInt(str));
        }

        //implementation of algorithm

        //sort by day
        bisDay(intDays, packs);
        System.out.println("--------------------------");
        ArrayList<Truck> trucks = new ArrayList<>();
        trucks.add(new SmallTruck());
        assignPacks(trucks, packs);



        //output trucks and content of trucks into deliveries.txt
        try {
            BufferedWriter br = new BufferedWriter(new FileWriter(deliveryFile));
            br.write("LIST OF TRUCKS WITH PACKAGE INFO\n");
            br.write("================================\n");
            br.newLine();
            br.newLine();
            int truckIndex = 1;
            for(Truck truck : trucks) {
                br.write("Truck " + truckIndex + ": "+ truck.toString() + "\n");
                br.write("-------------------\n");
                for(Pack pack : truck.getPackages()) {
                    br.write(pack.toString() + "\n");
                }
                br.newLine();
                truckIndex++;
            }
            br.close();
        } catch (IOException e) {
            e.printStackTrace();
        }


        //calculate and display truck hours
        int smallTruckCount = 0;
        int mediumTruckCount = 0;
        int largeTruckCount = 0;
        int totalTruckHours = 0;

        for(Truck truck : trucks) {
            if (truck instanceof SmallTruck) {
                smallTruckCount++;
                totalTruckHours += truck.totalTruckHours(); //Small Truck: 1 hour
            }else if (truck instanceof MediumTruck) {
                mediumTruckCount++;
                totalTruckHours += 2 * truck.totalTruckHours(); //Medium Truck: 2 hours
            }else if (truck instanceof LargeTruck) {
                largeTruckCount++;
                totalTruckHours += 3 * truck.totalTruckHours(); //Large Truck: 3 hours
            }
        }

        //display number of each type of truck and total truck hours to screen
        System.out.println("Number of Small Trucks: " + smallTruckCount);
        System.out.println("Number of Medium Trucks: " + mediumTruckCount);
        System.out.println("Number of Large Trucks: " + largeTruckCount);
        System.out.println("Total Truck Hours: " + totalTruckHours);

    }

    public static void bisDay(ArrayList<Integer> dayNum, ArrayList<Pack> packs) {
        for(int i = 0; i < dayNum.size(); i++){
            int temp = dayNum.get(i);
            Pack temps = packs.get(i);
            int low = 0;
            int high = i;
            while(low <= high){
                int mid = (low+high)/2;
                if(dayNum.get(mid) < temp){
                    low = mid + 1;
                }
                else {
                    high = mid - 1;
                }
            }
            for(int j=i-1; j>=low; j--){
                dayNum.set(j+1, dayNum.get(j));
                packs.set(j+1, packs.get(j));
            }
            dayNum.set(low, temp);
            packs.set(low, temps);
        }
    }

    public static int findNthOccurrence(String s, char ch, int n) {
        int occurrence = 0;
        for(int i=0; i<s.length(); i++){
            if(s.charAt(i) ==ch){
                occurrence +=1;
            }
            if(occurrence == n){
                return i;
            }
        }
        return -1;
    }

    public static Pack parsePackData(String[] data) {
        String receiver = data[1];
        String zone = data[2];
        String[] dateData = data[3].split("/");
        Date deliveryDate = new Date(Integer.parseInt(dateData[0]),
                                    Integer.parseInt(dateData[1]),
                                        Integer.parseInt(dateData[2]));
        int weight = Integer.parseInt(data[4]);
        int volume = Integer.parseInt(data[5]);

        if (data[0].equals("S")) {
            int time = Integer.parseInt(data[6]);
            return new SpecPack(receiver, zone, deliveryDate, weight, volume, time);
        }
        else {
            return new Pack(receiver, zone, deliveryDate, weight, volume);
        }
    }

    public static void assignPacks(ArrayList<Truck> truckList, ArrayList<Pack> packList) {
        List<Pack> assignedPackages = new ArrayList<>();
        boolean assigned;

        for (Pack packageItem : packList) {
            if (assignedPackages.contains(packageItem)) {
                break;
            }
            else {
                assigned = false;
                for (Truck truck : truckList) {
                    int totalVolume = truckList.stream().mapToInt(Truck::totalVolume).sum();
                    int totalWeight = truckList.stream().mapToInt(Truck::totalWeight).sum();

                    if (totalVolume + packageItem.getVolume() < truck.getMaxVolume() ||
                        totalWeight + packageItem.getWeight() < truck.getMaxWeight()) {
                        truck.addPack(packageItem);
                        assigned = true;
                        assignedPackages.add(packageItem);
                        break;
                    }
                }
                if(!assigned) {
                    LargeTruck newTruck = new LargeTruck();
                    truckList.add(newTruck);
                    newTruck.addPack(packageItem);
                    assignedPackages.add(packageItem);
                }
            }
        }
    }

}

File: Python/package_delivery_system.py
from Date import Date
from Pack import Pack
from SpecialPack import SpecialPack
from Truck import Truck
from small_truck import SmallTruck
from medium_truck import MediumTruck
from large_truck import LargeTruck


def bis_day(day_num, p):
    for i in range(len(day_num)):
        temp = day_num[i]
        temps = p[i]
        low = 0
        high = i
        while low <= high:
            mid = (low + high) // 2
            if day_num[mid] < temp:
                low = mid + 1
            else:
                high = mid - 1
        for j in range(i - 1, low - 1, -1):
            day_num[j+1] = day_num[j]
            p[j + 1] = p[j]
        day_num[low] = temp
        p[low] = temps

def find_nth_occurrence(s, ch, n):
    occurrence = 0
    for i in range(len(s)):
        if s[i] == ch:
            occurrence += 1
        if occurrence == n:
            return i
    return -1

def parse_pack_data(data):
    if data[0] == "S":
        receiver = data[1]
        zone = data[2]
        date_data = data[3].split("/")
        delivery_date = Date(int(date_data[0]), int(date_data[1]), int(date_data[2]))
        weight = int(data[4])
        volume = int(data[5])
        time = int(data[6])
        return SpecialPack(receiver, zone, delivery_date, weight, volume, time)
    else:
        receiver = data[1]
        zone = data[2]
        date_data = data[3].split("/")
        delivery_date = Date(int(date_data[0]), int(date_data[1]), int(date_data[2]))
        weight = int(data[4])
        volume = int(data[5])
        return Pack(receiver, zone, delivery_date, weight, volume)

def assign_packs(truck_list, pack_list):
    assigned_packages = []
    for package in pack_list:
        if package in assigned_packages:
            continue
        else:
            assigned = False
            for truck_item in truck_list:
                totalVolume = sum(truck_item.total_volume() for truck_item in truck_list)
                totalWeight = sum(truck_item.total_weight() for truck_item in truck_list)
                if totalVolume + package.get_volume() < truck_item.MAX_VOLUME or \
                        totalWeight + package.get_weight() < truck_item.MAX_WEIGHT:
                    truck_item.add_pack(package)
                    assigned = True
                    assigned_packages.append(package)
                    break
            if not assigned:
                new_truck = LargeTruck()
                truck_list.append(new_truck)
                new_truck.add_pack(package)
                assigned_packages.append(package)

#getting all text files
packages_file = input("enter packages file location: ")

#creating/managing collection of package objects

delivery_file = "deliveries.txt"
error_log = "log.txt"
#count how many packages are in the text file

number_of_packages = 0
#read each line in the text file to see how many packages there are
#IOError will write to a text file named log.txt

try:
    with open(packages_file, "r") as pack_file:
        for line in pack_file:
            number_of_packages += 1
except IOError as e:
    with open(error_log, "a") as log_file:
        log_file.write("IOException\n")
#make a list called list and a list named packs
#list will be used to read each line in the text file and store it
#each line in the text file will be converted to pack objects and then will be stored in packs
list = []
packs = []
#read lines from the file and store them in the list


with open(packages_file, "r") as pack_file:
    list = pack_file.readlines()
#convert each line in the list into a pack object and store in a pack list
#IOError will write to log.txt

for i in range(number_of_packages):
    if list[i][0] == "S" or list[i][0] == "R" and list[i][1] == ",":
        data = list[i].split(",")
        packs.append(parse_pack_data(data))
    else:
        with open(error_log, 'a') as writer:
            writer.write("IOException\n")
#make a string list of pack days
#this is done in order to sort the packs by day
#string list of packs
string_packs = []

for i in packs:
    string_packs.append(i.get_delivery_date().__str__())
#string list of days
days = []

for i in string_packs:
    index_slash = i.find("/")
    days.append(i[index_slash+1:index_slash+3])

"""
#just to check the days for the pack objects
for i in days:
    print(i)

"""
#string list to int list
int_days = []

for i in days:
    int_days.append(int(i))

#implementing algorithm
#sort packs by day by using binary insertion
#make a list of trucks and package ids
#we need a list of package ids to see which ones are already in a truck

bis_day(int_days, packs)
trucks = [SmallTruck()]
assign_packs(trucks, packs)


#output trucks and content of trucks into deliveries.txt
with open(delivery_file, "w") as output:
    output.write("LIST OF TRUCKS WITH PACKAGE INFO\n")
    output.write("================================\n")
    output.write("\n")
    output.write("\n")
    for i, truck in enumerate(trucks, start=1):
        output.write(f"Truck {i}: " + str(truck) +"\n")
        output.write("------------------\n")
        for j in truck.get_packages():
            output.write(str(j) + "\n")
        output.write("\n")

small_truck_count = 0
medium_truck_count = 0
large_truck_count = 0
total_truck_hours = 0

#calculate truck hours
for truck in trucks:
    if isinstance(truck, SmallTruck):
        small_truck_count += 1
        total_truck_hours += 1 * truck.total_truck_hours()  # Small truck: 1 hour
    elif isinstance(truck, MediumTruck):
        medium_truck_count += 1
        total_truck_hours += 2 * truck.total_truck_hours()  # Medium truck: 2 hours
    elif isinstance(truck, LargeTruck):
        large_truck_count += 1
        total_truck_hours += 3 * truck.total_truck_hours()  # Large truck: 3 hours

# Display results to the screen
print(f"Number of Small Trucks: {small_truck_count}")
print(f"Number of Medium Trucks: {medium_truck_count}")
print(f"Number of Large Trucks: {large_truck_count}")
print(f"Total Truck Hours: {total_truck_hours}")
